### DISCUSSION TITLE
SystemError is lacking from Node typing [node]

### DISCUSSION THREAD

**Discussion by wvhulle** [DefinitelyTyped/DefinitelyTyped#64508] [Issues with a @types package] [answered]
Hi, I could not find any typings for the built-in node SystemError class.

[a comment by an automated account is not shown]

**Comment by Semigradsky** [comment]
What the use case?

**Reply by wvhulle** [reply]
I wanted to do "instanceof SystemError" but Typescript was complaining that SystemError is undefined and I couldn't find it anywhere to import. Apparently the class itself is not exported by Node. This kind of error has some special fields, so it could be useful to have types for it.

**Reply by Semigradsky** [reply]
You can't use `instanceof SystemError` if the class itself is not exported by Node.

The node-way to checking what error it is - checking `error.code` and/or `error.errno`. The typings already have common `ErrnoException` class with these fields.

**Reply by Semigradsky** [reply]
Exporting `SystemError` interface might be useful for using in your code like `(error as SystemError).address`. But you can't use `instanceof SystemError`.

**Reply by edwardgalligan** [reply]
`ErrNoException` works fine for `code` specifically but it only supports 4 of the 9 properties on NodeJS's `SystemError`.

For example this use-case isn't supported:

```
try {
  functionCall();
} catch (e) {
  if (e.info) { // this line is invalid in Typescript using `strict: true`
    console.log({ info: e.info });
  }
}
```

**Reply by sdavids** [reply]
Here is a another use case where it would be useful:

```js
httpServer.listen(port);

httpServer.on('error', (err) => {
  if (err.syscall === 'listen') {
    switch (err.code) {
      case 'EACCES':
        console.error('Port requires elevated privileges');
        process.exit(77); // EX_NOPERM
        break;
      case 'EADDRINUSE':
        console.error('Port is already in use');
        process.exit(75); // EX_TEMPFAIL
        break;
      default:
      // just log
    }
  }

  console.error(err);
});
```

**Comment by wvhulle** [comment, marked as answer]
My solution was to switch to Rust. Typescript is hopeless.
